chainer_chemistry/models/gat.py

```diff
@@ -0,0 +1,169 @@
+                  for _ in range(n_message_layer)]
+            )
+            self.attenstion_layers = chainer.ChainList(
+                *[GraphLinear(hidden_dim * 2, 1)
+                  for _ in range(n_message_layer)]
+            )
+            # self.update_layer = links.GRU(2 * hidden_dim, hidden_dim)
+            # Readout
+            self.i_layers = chainer.ChainList(
+                *[GraphLinear(2 * hidden_dim, out_dim)
+                    for _ in range(n_readout_layer)]
+            )
+            self.j_layers = chainer.ChainList(
+                *[GraphLinear(hidden_dim, out_dim)
+                    for _ in range(n_readout_layer)]
+            )
+        self.out_dim = out_dim
+        self.heads = heads
+        self.hidden_dim = hidden_dim
+        self.n_layers = n_layers
+        self.concat_hidden = concat_hidden
+        self.weight_tying = weight_tying
+        self.negative_slope = negative_slope
+
+    def update(self, h, adj, step=0):
+        xp = self.xp
+        # (minibatch, atom, channel)
+        mb, atom, ch = h.shape
+        # (minibatch, atom, heads * out_dim)
+        h = self.message_layers[step](h)
+        # (minibatch, atom, heads, out_dim)
+        h = functions.reshape(h, (mb, atom, self.heads, self.hidden_dim))
+
+        # concat all pairs of atom
+        # (minibatch, 1, atom, heads, out_dim)
+        h_i = functions.expand_dims(h, axis=1)
+        # (minibatch, atom, atom, heads, out_dim)
+        h_i = functions.broadcast_to(h_i, (mb, atom, atom,
+                                           self.heads, self.hidden_dim))
+        device_id = cuda.get_device_from_array(h_i.array).id
+        h_j = functions.copy(h_i, device_id)
+        h_j = functions.transpose(h_j, (0, 2, 1, 3, 4))
+
+        # (minibatch, atom, atom, heads, out_dim * 2)
+        e = functions.concat([h_i, h_j], axis=4)
+
+        # (minibatch, heads, atom, atom, out_dim * 2)
+        e = functions.transpose(e, (0, 3, 1, 2, 4))
+        # (minibatch * heads, atom * atom, out_dim * 2)
+        e = functions.reshape(e, (mb * self.heads, atom * atom,
+                                  self.hidden_dim * 2))
+        # (minibatch * heads, atom * atom, 1)
+        e = self.attenstion_layers[step](e)
+
+        # (minibatch, heads, atom, atom)
+        e = functions.reshape(e, (mb, self.heads, atom, atom))
+        # (heads, minibatch, atom, atom)
+        e = functions.transpose(e, (1, 0, 2, 3))
+        e = functions.leaky_relu(e)
+        # z = functions.reshape(z, (self.heads, mb, atom, atom))
+
+        cond = adj.array.astype(xp.bool)
+        cond = xp.broadcast_to(cond, e.array.shape)
+        # TODO(mottodora): find better way to ignore non connected
+        e = functions.where(cond, e,
+                            xp.broadcast_to(xp.array(-10000), e.array.shape)
+                            .astype(xp.float32))
+        # (heads, minibatch, atom, atom)
+        alpha = functions.softmax(e)
```

Comment:
softmax axis is 1, do you want to apply along minibatch??

Reply:
This is a mistake. Thank you!